stdlib/src/sys/ffi/c/libc.mojo

```diff
@@ -0,0 +1,2498 @@
+
+    @always_inline
+    fn fgetc(self, stream: UnsafePointer[FILE]) -> C.int:
+        """Libc POSIX `fgetc` function.
+
+        Args:
+            stream: A pointer to a stream.
+
+        Returns:
+            The next byte from the input stream pointed to by stream. If the
+            end-of-file indicator for the stream is set, or if the stream is at
+            end-of-file, the end-of-file indicator for the stream shall be set
+            and `fgetc()` shall return EOF. If a read error occurs, the error
+            indicator for the stream shall be set, fgetc() shall return `EOF`,
+            and shall set `errno` to indicate the error.
+
+        Notes:
+            [Reference](https://man7.org/linux/man-pages/man3/fgetc.3p.html).
+            Fn signature: `int fgetc(FILE *stream)`.
+        """
+
+        @parameter
+        if static:
+            return external_call["fgetc", C.int](stream)
+        else:
+            return self._lib.value().call["fgetc", C.int](stream)
+
+    @always_inline
+    fn fgets(
+        self, s: UnsafePointer[C.char], n: C.int, stream: UnsafePointer[FILE]
+    ) -> UnsafePointer[C.char]:
+        """Libc POSIX `fgets` function.
+
+        Args:
+            s: A pointer to a buffer to store the read string.
+            n: The maximum number of characters to read.
+            stream: A pointer to a stream.
+
+        Returns:
+            Upon successful completion, fgets() shall return s. If the stream is
+            at end-of-file, the end-of-file indicator for the stream shall be
+            set and `fgets()` shall return a null pointer. If a read error
+            occurs, the error indicator for the stream shall be set, `fgets()`
+            shall return a null pointer, and shall set `errno` to indicate the
+            error.
+
+        Notes:
+            [Reference](https://man7.org/linux/man-pages/man3/fgets.3p.html).
+            Fn signature: `char *fgets(char *restrict s, int n,
+                FILE *restrict stream)`.
+        """
+
+        @parameter
+        if static:
+            return external_call["fgets", UnsafePointer[C.char]](s, n, stream)
+        else:
+            return self._lib.value().call["fgets", UnsafePointer[C.char]](
+                s, n, stream
+            )
+
+    @always_inline
+    fn printf(
+        self,
+        format: UnsafePointer[C.char],
+        args: VariadicPack[element_trait=AnyType],
+    ) -> C.int:
+        """Libc POSIX `printf` function.
+
+        Args:
+            format: The format string.
+            args: The arguments to be added into the format string.
+
+        Returns:
+            The number of bytes printed excluding the terminating null byte.
+            Otherwise a negative value and `errno` is set.
+
+        Notes:
+            [Reference](https://man7.org/linux/man-pages/man3/printf.3p.html).
+        """
+        var a = args.get_loaded_kgen_pack()
+
+        @parameter
+        if is_nvidia_gpu():
```

Comment:
Could just `is_gpu()` work?

Reply:
I saw this workaround in a stdlib function, I don't know if it's some NVIDIA particular thing or if it's that the API for GPUs needs to be different :man_shrugging:. This `vprintf` signature is actually non standard AFAIK since it doesn't use the [`va_list`](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdarg.h.html) type, but it could also be that the underlying `VariadicPack` uses such a structure internally and when a Pointer traverses though it it works fine, when I tested `vprintf` on Linux with the same setup it did not work :man_shrugging:.